enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -566,6 +629,110 @@ func (p *PebbleCache) updateAtime(key filestore.PebbleKey) error {
 	return db.Set(keyBytes, protoBytes, &pebble.WriteOptions{Sync: false})
 }
 
+func (p *PebbleCache) migrateData(quitChan chan struct{}) error {
+	if len(p.migrators) == 0 {
+		log.Debugf("No migrations necessary")
+		return nil
+	}
+
+	limiter := rate.NewLimiter(rate.Limit(*migrationQPSLimit), 1)
+	db, err := p.leaser.DB()
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	iter := db.NewIter(&pebble.IterOptions{
+		LowerBound: keys.MinByte,
+		UpperBound: keys.MaxByte,
+	})
+	defer iter.Close()
+
+	minVersion := p.maxDatabaseVersion()
+	maxVersion := p.minDatabaseVersion()
+	migrationStart := time.Now()
+	keysSeen := 0
+	keysMigrated := 0
+	lastStatusUpdate := time.Now()
+
+	for iter.First(); iter.Valid(); iter.Next() {
+		if bytes.HasPrefix(iter.Key(), SystemKeyPrefix) {
+			continue
+		}
+		keysSeen += 1
+
+		select {
+		case <-quitChan:
+			return nil
+		default:
+		}
+		_ = limiter.Wait(p.env.GetServerContext())
+
+		if time.Since(lastStatusUpdate) > 10*time.Second {
+			log.Infof("Pebble Cache: data migration progress: saw %d keys, migrated %d to version: %d in %s", keysSeen, keysMigrated, maxVersion, time.Since(migrationStart))
+			lastStatusUpdate = time.Now()
+		}
+		var key filestore.PebbleKey
+		version, err := key.FromBytes(iter.Key())
+		if err != nil {
+			return err
+		}
+		oldVersion := version
+		valBytes := iter.Value()
+
+		for _, migrator := range p.migrators {
+			// If this key was already migrated, skip it.
+			if version >= migrator.ToVersion() {
+				continue
+			}
+
+			// If this key does not match this migrator's
+			// "FromVersion", then the migrators that should have
+			// run before this one did not, and something is wrong.
+			// Bail out.
+			if version != migrator.FromVersion() {
+				return status.FailedPreconditionErrorf("Migrator %+v cannot migrate key from version %d", migrator, version)
+			}
+
+			valBytes = migrator.Migrate(valBytes)
+			version = migrator.ToVersion()
+			keysMigrated += 1
+
+			if version > maxVersion {
+				maxVersion = version
```

Comment:
This logic seems off. It looks like you can end up with the wrong versions at the end of the migration. Let's say you set write version to v2 and have two migrators, v0->v1 and v1->v2. 
After the loop, you will have minVersion=v1 and maxVersion=v2 where's what you want is minVersion=v2 since there should be no v1 keys at that point.